### PR TITLE
feat(#2746):tests show that file verification don't work correctly

### PR DESCRIPTION
**Pull request by Yanich96**
New tests was added in OptCached. They show that file verification form cache don't work correctly. This is the reason of fails OptimizeMojoTest and ShakeMojoTest.

---

## PR-Codex overview
This PR focuses on fixing the caching issue in the `OptCached` class by using checksum instead of time. 

### Detailed summary
- Changes the caching mechanism to use checksum instead of time.
- Adds tests to verify the correctness of fetching from the cache.
- Updates the documentation to reflect the changes made.

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by Yanich96** [comment]
@volodya-lombrozo check please

File: eo-maven-plugin/src/main/java/org/eolang/maven/optimization/OptCached.java [hunk 1]
```diff
@@ -41,9 +41,12 @@
  * Returns already optimized XML if it's found in the cache.
  *
  * @since 0.28.11
- * @todo #2674:30min The function {@code  OptCached.contains(final XML xml)}
- *  isn't work properly. This function compares caching and compilation times,
- *  which may lead to erroneous results. We need to fix this.
+ * @todo #2746:30min The tests
+ *  {@link org.eolang.maven.optimization.OptCachedTest#returnsFromCacheCorrectProgram(Path path)}
```
Comment (volodya-lombrozo) [inline eo-maven-plugin/src/main/java/org/eolang/maven/optimization/OptCached.java:45]: @Yanich96 What do you think about the following description?
```java
// @todo #2746: 30min Use checksum, not time.
//  The following tests show that fetching from the cache doesn't work correctly:
//   - {@link org.eolang.maven.optimization.OptCachedTest#returnsFromCacheCorrectProgram(Path path)}
//   - {@link org.eolang.maven.optimization.OptCachedTest#returnsFromCacheButTimesSaveAndExecuteDifferent(Path path)}
//  Need to fix file validation from the cache: using checksum, but not time.
//  Don't forget to enable the tests.
```
Fixed grammar mistakes and the name of the test.

File: eo-maven-plugin/src/test/java/org/eolang/maven/optimization/OptCachedTest.java [hunk 2]
```diff
@@ -74,6 +74,44 @@ void returnsFromCacheIfXmlAlreadyInCache(@TempDir final Path tmp) throws IOExcep
         );
     }
 
+    @Disabled
+    @Test
+    void returnsFromCacheButTimesSaveAndExecuteDifferent(@TempDir final Path tmp)
+        throws IOException, InterruptedException {
+        final XML program = OptCachedTest.program(ZonedDateTime.now());
+        Thread.sleep(70_000);
```
Comment (volodya-lombrozo) [inline eo-maven-plugin/src/test/java/org/eolang/maven/optimization/OptCachedTest.java:82]: @Yanich96 It is a ["Wait and See"](https://stackoverflow.com/a/339339/10423604) anti-pattern

File: eo-maven-plugin/src/test/java/org/eolang/maven/optimization/OptCachedTest.java [hunk 3]
```diff
@@ -158,4 +196,23 @@ private static XML program(final ZonedDateTime time) {
...
         );
     }
+
+    /**
+     * Generates EO program for tests with specified time and content.
+     * @param time Time.
+     * @param something String.
+     * @return XML representation of program.
+     */
+    private static XML programWithSomethings(final ZonedDateTime time, final String something) {
```
Comment (volodya-lombrozo) [inline eo-maven-plugin/src/test/java/org/eolang/maven/optimization/OptCachedTest.java:206]: @Yanich96 Maybe we can overload the already existed `private static XML program(final ZonedDateTime time)` method? In this case we won't need to repeat the same code twice.

File: eo-maven-plugin/src/test/java/org/eolang/maven/optimization/OptCachedTest.java [hunk 4]
```diff
@@ -74,6 +74,44 @@ void returnsFromCacheIfXmlAlreadyInCache(@TempDir final Path tmp) throws IOExcep
...
 
+    @Disabled
+    @Test
+    void returnsFromCacheButTimesSaveAndExecuteDifferent(@TempDir final Path tmp)
+        throws IOException, InterruptedException {
+        final XML program = OptCachedTest.program(ZonedDateTime.now());
+        Thread.sleep(70_000);
+        OptCachedTest.save(tmp, program);
+        MatcherAssert.assertThat(
+            "We expected that the program will be returned from the cache",
```
Comment (volodya-lombrozo) [inline eo-maven-plugin/src/test/java/org/eolang/maven/optimization/OptCachedTest.java:85]: @Yanich96, this message doesn't provide useful information for a developer. Perhaps you could explain what you specifically expect in this test case and add more context?

File: eo-maven-plugin/src/test/java/org/eolang/maven/optimization/OptCachedTest.java [hunk 5]
```diff
@@ -74,6 +74,44 @@ void returnsFromCacheIfXmlAlreadyInCache(@TempDir final Path tmp) throws IOExcep
...
+
+    @Disabled
+    @Test
+    void returnsFromCacheCorrectProgram(@TempDir final Path tmp)
+        throws IOException, InterruptedException {
+        XML program = OptCachedTest.programWithSomethings(ZonedDateTime.now(), "first program");
+        OptCachedTest.save(tmp, program);
+        program = OptCachedTest.programWithSomethings(ZonedDateTime.now(), "second program");
+        MatcherAssert.assertThat(
+            "We expected that the correct program will be returned from the cache",
```
Comment (volodya-lombrozo) [inline eo-maven-plugin/src/test/java/org/eolang/maven/optimization/OptCachedTest.java:104]: @Yanich96, this message doesn't provide useful information for a developer. Perhaps you could explain what you specifically expect in this test case and add more context?

**Review by volodya-lombrozo — CHANGES_REQUESTED** [review]
@Yanich96 Could you [link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) the PR with the issue you are tying to solve, please?

File: eo-maven-plugin/src/test/java/org/eolang/maven/optimization/OptCachedTest.java (hunk 3 above)
Comment (Yanich96) [reply eo-maven-plugin/src/test/java/org/eolang/maven/optimization/OptCachedTest.java:206]: @volodya-lombrozo I think existed method doesn't show needed results. I did the same program with a different context, so I can show that time verification is not reliable

Comment (volodya-lombrozo) [reply eo-maven-plugin/src/test/java/org/eolang/maven/optimization/OptCachedTest.java:206]: @Yanich96 I understood your intentions here. I just suggest to apply [method overloading](https://en.wikipedia.org/wiki/Function_overloading):
```java
    private static XML program(final ZonedDateTime time) {
        return OptCachedTest.program(time, "same");
    }

    private static XML program(final ZonedDateTime time, final String content) {
        return new XMLDocument(
            new Xembler(
                new Directives()
                    .add("program")
                    .attr("name", "main")
                    .attr("time", time.format(DateTimeFormatter.ISO_INSTANT))
                    .attr("something", content)
                    .up()
            ).xmlQuietly()
        );
    }
```

File: eo-maven-plugin/src/test/java/org/eolang/maven/optimization/OptCachedTest.java (hunk 4 above)
Comment (Yanich96) [reply eo-maven-plugin/src/test/java/org/eolang/maven/optimization/OptCachedTest.java:85]: @volodya-lombrozo is the message normal?
 "We expected that the program will be returned from the cache, " +
                "even if compilation and saving times are different"

File: eo-maven-plugin/src/test/java/org/eolang/maven/optimization/OptCachedTest.java (hunk 5 above)
Comment (Yanich96) [reply eo-maven-plugin/src/test/java/org/eolang/maven/optimization/OptCachedTest.java:104]: @volodya-lombrozo is the message normal?
"We expected that the correct program will be returned from the cache, " +
 "but time verification is not reliable"

Comment (volodya-lombrozo) [reply eo-maven-plugin/src/test/java/org/eolang/maven/optimization/OptCachedTest.java:104]: @Yanich96 Yes, I like it.

File: eo-maven-plugin/src/test/java/org/eolang/maven/optimization/OptCachedTest.java (hunk 4 above)
Comment (volodya-lombrozo) [reply eo-maven-plugin/src/test/java/org/eolang/maven/optimization/OptCachedTest.java:85]: @Yanich96 Yes, I like it.

**Comment by Yanich96** [comment]
@volodya-lombrozo check please

**Review by volodya-lombrozo — CHANGES_REQUESTED** [review]
It's important for traceability:
> @Yanich96 Could you [link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) the PR with the issue you are tying to solve, please?

**Comment by Yanich96** [comment]
@volodya-lombrozo check please

**Review by volodya-lombrozo — APPROVED** [review]
@Yanich96 Looks good. Thanks.
@yegor256 Could you merge that changes, please?

**Comment by Yanich96** [comment]
@yegor256 could you merge, please?

**Comment by yegor256** [comment]
@rultor merge

**Pull request merged**